### PR TITLE
passphrase encryption - Closes #325

### LINKED ISSUE #325: AES256-cbc encryption / decryption
For the forging passphrase

### PR DESCRIPTION
**Pull request by toschdev**
Closes #325 

Merge https://github.com/LiskHQ/lisk-js/pull/324 first

### CODE REVIEW

File: src/crypto/sign.js [hunk 1]
```diff
@@ -346,3 +346,42 @@ export function verifySecondSignature(transaction, publicKey) {
...
+ * @return {string}
+ */
+
+export function aesEncrypt(passphrase, password) {
+	const iv = crypto.randomBytes(16);
+	const passHash = crypto.createHash('sha256').update(password, 'utf8').digest();
+	const cipher = crypto.createCipheriv('aes-256-cbc', passHash, iv);
+	const encrypted = cipher.update(passphrase);
+	const encryptedString = Buffer.concat([encrypted, cipher.final()]);
+	const cipherText = ''.concat(iv.toString('hex'), '$', encryptedString.toString('hex'));
```
Comment (willclarktech) [inline src/crypto/sign.js]: Since the IV is always the same length we can save a character by skipping the `$` here.

File: src/crypto/sign.js [hunk 2]
```diff
@@ -346,3 +346,42 @@ export function verifySecondSignature(transaction, publicKey) {
...
+
+	return cipherText;
```
Comment (willclarktech) [inline src/crypto/sign.js]: Prefer concatenating buffers and returning a string right at the end.

File: src/crypto/sign.js [hunk 3]
```diff
@@ -346,3 +346,42 @@ export function verifySecondSignature(transaction, publicKey) {
...
+ */
+
+export function aesDecrypt(cipherText, password) {
+	const passHash = crypto.createHash('sha256').update(password, 'utf8').digest();
+	const encryptedParts = cipherText.split('$');
+	const iv = Buffer.from(encryptedParts.shift(), 'hex');
+	const encryptedPass = Buffer.from(encryptedParts.join('$'), 'hex');
+	const decipher = crypto.createDecipheriv('aes-256-cbc', Buffer.from(passHash), iv);
+	let decrypted = decipher.update(encryptedPass);
+	decrypted = Buffer.concat([decrypted, decipher.final()]);
```
Comment (willclarktech) [inline src/crypto/sign.js]: Prefer two different `const`s

File: test/crypto/sign.js [hunk 4]
```diff
@@ -315,5 +315,26 @@ ${defaultSignature}
...
 		});
+
+		describe('encrypting passphrase', () => {
+			const secretPassphrase = 'minute omit local rare sword knee banner pair rib museum shadow juice';
+			const password = 'myTotal53cr3t%&';
+			const encryptString = cryptoModule.aesEncrypt(secretPassphrase, password);
+
+			describe('#aesEncrypt', () => {
+				it('should encrypt a given secret with a password', () => {
+					(encryptString).should.be.ok();
```
Comment (willclarktech) [inline test/crypto/sign.js]: This test adds nothing over the next one.

File: test/crypto/sign.js [hunk 5]
```diff
@@ -315,5 +315,26 @@ ${defaultSignature}
...
+					(encryptString).should.containEql('$');
+				});
+			});
+
+			describe('#aesDecrypt', () => {
+				it('should decrypt a given cipher with a password', () => {
+					const decryptedString = cryptoModule.aesDecrypt(encryptString, password);
+					(decryptedString).should.be.eql(secretPassphrase);
+				});
+			});
```
Comment (willclarktech) [inline test/crypto/sign.js]: These tests are not atomic. (Although an additional integration test would make sense.)

File: test/crypto/sign.js [hunk 6]
```diff
@@ -315,5 +315,26 @@ ${defaultSignature}
...
+		describe('encrypting passphrase', () => {
+			const secretPassphrase = 'minute omit local rare sword knee banner pair rib museum shadow juice';
+			const password = 'myTotal53cr3t%&';
+			const encryptString = cryptoModule.aesEncrypt(secretPassphrase, password);
+
+			describe('#aesEncrypt', () => {
+				it('should encrypt a given secret with a password', () => {
+					(encryptString).should.be.ok();
+					(encryptString).should.be.type('string');
+					(encryptString).should.containEql('$');
```
Comment (willclarktech) [inline test/crypto/sign.js]: What should the output be?

File: test/crypto/sign.js [hunk 7]
```diff
@@ -315,5 +315,26 @@ ${defaultSignature}
 				(multiSignature).should.be.eql(expectedMultiSignature);
 			});
 		});
+
+		describe('encrypting passphrase', () => {
```
Comment (willclarktech) [inline test/crypto/sign.js]: `AES-256-CBC encryption using a password`

**Review by willclarktech — CHANGES_REQUESTED** [review]

File: src/crypto/sign.js (hunk 1 above)
Comment (Tosch110) [reply src/crypto/sign.js]: I like the delimiter. It leaves room to change the length of the IV at a certain point and gives a clear structure

File: src/crypto/sign.js [hunk 8]
```diff
@@ -346,3 +346,42 @@ export function verifySecondSignature(transaction, publicKey) {
...
+ * @method aesEncrypt
+ * @param passphrase
+ * @param password
+ *
+ * @return {string}
+ */
+
+export function aesEncrypt(passphrase, password) {
+	const iv = crypto.randomBytes(16);
+	const passHash = crypto.createHash('sha256').update(password, 'utf8').digest();
```
[the remark this answers is not in the document]
Comment (Tosch110) [reply src/crypto/sign.js]: createCipheriv from nodejs expects a specific length for the password. With a hash we get sure about the length

**Pull request closed without merging**